internal/proxy/middleware/attributes/attributes.go

```diff
@@ -0,0 +1,174 @@
+		requestAttributes["prefix"] = rule.Backend.Prefix
+	}
+
+	for res, attr := range config.Attributes {
+		_ = res
+
+		switch attr.Type {
+		case middleware.AttributeTypeGRPCPayload:
+			// check if grpc request
+			if !strings.HasPrefix(req.Header.Get("Content-Type"), "application/grpc") {
+				a.log.Error("middleware: not a grpc request", "attr", attr)
+				a.notAllowed(rw)
+				return
+			}
+
+			// TODO: we can optimise this by parsing all field at once
+			payloadField, err := body_extractor.GRPCPayloadHandler{}.Extract(&req.Body, attr.Index)
+			if err != nil {
+				a.log.Error("middleware: failed to parse grpc payload", "err", err)
+				return
+			}
+
+			requestAttributes[res] = payloadField
+			a.log.Info("middleware: extracted", "field", payloadField, "attr", attr)
+
+		case middleware.AttributeTypeJSONPayload:
+			if attr.Key == "" {
+				a.log.Error("middleware: payload key field empty")
+				a.notAllowed(rw)
+				return
+			}
+			payloadField, err := body_extractor.JSONPayloadHandler{}.Extract(&req.Body, attr.Key)
+			if err != nil {
+				a.log.Error("middleware: failed to parse grpc payload", "err", err)
+				a.notAllowed(rw)
+				return
+			}
+
+			requestAttributes[res] = payloadField
+			a.log.Info("middleware: extracted", "field", payloadField, "attr", attr)
+
+		case middleware.AttributeTypeHeader:
+			if attr.Key == "" {
+				a.log.Error("middleware: header key field empty")
+				a.notAllowed(rw)
+				return
+			}
+			headerAttr := req.Header.Get(attr.Key)
+			if headerAttr == "" {
+				a.log.Error(fmt.Sprintf("middleware: header %s is empty", attr.Key))
+				a.notAllowed(rw)
+				return
+			}
+
+			requestAttributes[res] = headerAttr
+			a.log.Info("middleware: extracted", "field", headerAttr, "attr", attr)
+
+		case middleware.AttributeTypeQuery:
+			if attr.Key == "" {
+				a.log.Error("middleware: query key field empty")
+				a.notAllowed(rw)
+				return
+			}
+			queryAttr := req.URL.Query().Get(attr.Key)
+			if queryAttr == "" {
+				a.log.Error(fmt.Sprintf("middleware: query %s is empty", attr.Key))
+				a.notAllowed(rw)
+				return
+			}
+
+			requestAttributes[res] = queryAttr
+			a.log.Info("middleware: extracted", "field", queryAttr, "attr", attr)
+
+		case middleware.AttributeTypeConstant:
+			if attr.Value == "" {
+				a.log.Error("middleware: constant value empty")
+				a.notAllowed(rw)
+				return
+			}
+
+			requestAttributes[res] = attr.Value
+			a.log.Info("middleware: extracted", "constant_key", res, "attr", requestAttributes[res])
+
+		default:
+			a.log.Error("middleware: unknown attribute type", "attr", attr)
+			a.notAllowed(rw)
+			return
+		}
+	}
+
+	paramMap, mapExists := middleware.ExtractPathParams(req)
+	if !mapExists {
+		a.log.Error("middleware: path param map doesn't exist")
+		a.notAllowed(rw)
+		return
+	}
+
+	for key, value := range paramMap {
+		requestAttributes[key] = value
+	}
+
+	attributesContext := context.WithValue(req.Context(), "requestAttributes", requestAttributes)
```

Comment:
do not use primitive string value as key in contexts. Refer to my other comment. the snippet provided in my other comment can be added into this package and used as `attributesContext := InjectAttribs(req.Context(), requestAttributes)`.